### PR TITLE
[libc] implement cached process/thread identity

### PR DESCRIPTION
**Pull request by SchrodingerZhu**

### CODE REVIEW

File: libc/spec/posix.td [hunk 1]
```diff
@@ -567,16 +572,6 @@ def POSIX : StandardSpec<"POSIX"> {
           RetValSpec<IntType>,
           [ArgSpec<ConstCharPtr>]
         >,
-        FunctionSpec<
```
Comment (michaelrj-google) [inline libc/spec/posix.td:570]: why were these deleted?

File: libc/docs/dev/undefined_behavior.rst [hunk 2]
```diff
@@ -89,3 +89,24 @@ The C23 standard states that if the value of the ``rnd`` argument of the
 the value of a math rounding direction macro, the direction of rounding is
 unspecified. LLVM's libc chooses to use the ``FP_INT_TONEAREST`` rounding
 direction in this case.
+
+Cached ``getpid/gettid``
+------------------------
+Since version ``2.25``, glibc removes its cache mechanism for ``getpid/gettid``
```
Comment (michaelrj-google) [inline libc/docs/dev/undefined_behavior.rst:95]: given the potential downsides, would it be possible to create an option to turn this caching on and off? If you do, then it would be good to mention that flag here.

File: libc/spec/posix.td (hunk 1 above)
Comment (SchrodingerZhu) [reply libc/spec/posix.td:570]: They were repeated somehow with different types. There are already definitions using `PidT` in the above.

File: libc/src/__support/OSUtil/pid.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+#include "src/__support/macros/attributes.h"
+#include "src/__support/macros/optimization.h"
+namespace LIBC_NAMESPACE {
+
+class ProcessIdentity {
+  static pid_t cache;
+  static pid_t get_uncached();
+
+public:
+  LIBC_INLINE static void invalidate_cache() { cache = -1; }
```
Comment (SchrodingerZhu) [inline libc/src/__support/OSUtil/pid.h:21]: When multiple threads want to fork in the same time, is there a race condition happening here?

Comment (SchrodingerZhu) [reply libc/src/__support/OSUtil/pid.h:21]: Ok, I believe this implementation of PID cache is wrong. I may remove the PID cache for now.

File: libc/src/unistd/linux/fork.cpp [hunk 4]
```diff
@@ -24,29 +25,41 @@ namespace LIBC_NAMESPACE {
...
 
   if (ret < 0) {
     // Error case, a child process was not created.
     libc_errno = static_cast<int>(-ret);
     return -1;
   }
 
+  // Common
+  // Refresh tid/pid cache after fork
+  self.refresh_tid();
```
Comment (QuarticCat) [inline libc/src/unistd/linux/fork.cpp:51]: Will the tid of parent process change after forking? If not, we can probably avoid syscall in the parent process here.

File: libc/src/__support/OSUtil/pid.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  LIBC_INLINE static void end_fork() { fork_inflight = false; }
+  LIBC_INLINE static void refresh_cache() { cache = get_uncached(); }
+  LIBC_INLINE static pid_t get() {
+#if LIBC_COPT_ENABLE_PID_CACHE
+    if (LIBC_UNLIKELY(fork_inflight))
+      return get_uncached();
+    return cache;
+#else
+    return get_uncached();
+#endif
```
Comment (nickdesaulniers) [inline libc/src/__support/OSUtil/pid.h:37]:
```suggestion
#if LIBC_COPT_ENABLE_PID_CACHE
    if (LIBC_LIKELY(!fork_inflight))
      return cache;
#endif
    return get_uncached();
```

File: libc/src/__support/threads/linux/thread.cpp [hunk 6]
```diff
@@ -517,4 +517,6 @@ void thread_exit(ThreadReturnValue retval, ThreadStyle style) {
   __builtin_unreachable();
 }
 
+pid_t Thread::get_uncached_tid() { return syscall_impl<pid_t>(SYS_gettid); }
```
Comment (nickdesaulniers) [inline libc/src/__support/threads/linux/thread.cpp:520]: Can we clean up the number of places that we call `SYS_gettid` directly?  Looks like `gettid_inline` does this, too.

**Review by michaelrj-google — APPROVED** [review]
LGTM

**Review by QuarticCat — APPROVED** [review]
LGTM

**Comment by SchrodingerZhu** [comment]
This PR somehow gets 50k+ commits diverged from main. I think it is a git bug but it makes me really hard to reconcile the changes.

**Comment by SchrodingerZhu** [comment]
superseded by https://github.com/llvm/llvm-project/pull/98989

**Comment by SchrodingerZhu** [comment]
@QuarticCat maybe this is a git hash conflict problem?
